src/frames.ts

```diff
@@ -46,6 +47,8 @@ export type GotoResult = {
 
 type ConsoleTagHandler = () => void;
 
+export type FunctionWithSource = (source: { context: BrowserContext, page: Page, frame: Frame}, ...args: any) => any;
```

Comment:
Having a single parameter with named accessors { source, args } could be better and more aligned with other APIs where we allow only single argument. Both approaches are non-ideal though, I'd look for a more elegant way before making this publicly available.

Reply:
Agreed, thought about the same thing